// y-postgresql/dist/pg-adapter.cjs
'use strict';

var pg = require('pg');
var format = require('pg-format');

const { Pool } = pg;
class PgAdapter {
    constructor(tableName, pool) {
        Object.defineProperty(this, "tableName", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "pool", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.tableName = tableName;
        this.pool = pool;
    }
    static async connect(connectionOptions, { tableName, useIndex }) {
        const pool = new Pool(connectionOptions);
        await pool.query('SELECT 1+1;');
        const tableExistsRes = await pool.query(`SELECT EXISTS (
                SELECT FROM pg_tables
                WHERE schemaname = 'public'
                AND tablename  = $1
            );`, [tableName]);
        const tableExists = tableExistsRes.rows[0].exists;
        if (!tableExists) {
            await pool.query(`
			DO $$
			BEGIN
				IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'ypga_version') THEN
					CREATE TYPE ypga_version AS ENUM ('v1', 'v1_sv');
				END IF;
			END
			$$;
			`);
            await pool.query(format(`
                CREATE TABLE %I (
                    id SERIAL PRIMARY KEY,
                    docname TEXT NOT NULL,
                    value BYTEA NOT NULL,
                    version ypga_version NOT NULL
                );`, tableName));
        }
        if (useIndex) {
            const indexDocNameExistsRes = await pool.query(`SELECT EXISTS (
                    SELECT FROM pg_indexes
                    WHERE tablename = $1
                    AND indexname = $1_docname_idx
                );`, [tableName]);
            const indexDocNameExists = indexDocNameExistsRes.rows[0].exists;
            if (!indexDocNameExists) {
                await pool.query(format(`
                CREATE INDEX %I_docname_idx ON %I (docname);`, tableName, tableName));
            }
        }
        return new PgAdapter(tableName, pool);
    }
    async findLatestDocumentId(docName) {
        const query = format(`
            SELECT id FROM %I
            WHERE docname = %L
            ORDER BY id DESC
            LIMIT 1;`, this.tableName, docName);
        const res = await this.pool.query(query);
        return res.rows[0]?.id ?? -1;
    }
    async insertUpdate(docName, value) {
        const bufferValue = Buffer.from(value);
        const query = format(`
            INSERT INTO %I (docname, value, version)
            VALUES (%L, %L, 'v1')
            RETURNING *;`, this.tableName, docName, bufferValue);
        const res = await this.pool.query(query);
        return res.rows[0];
    }
    async _getStateVector(docName) {
        const query = format(`
			SELECT value FROM %I
			WHERE docname = %L
			AND version = 'v1_sv'
			LIMIT 1;`, this.tableName, docName);
        const res = await this.pool.query(query);
        return res.rows[0];
    }
    async getStateVectorBuffer(docName) {
        const res = await this._getStateVector(docName);
        return res?.value;
    }
    async putStateVector(docName, value) {
        const bufferValue = Buffer.from(value);
        const sv = await this._getStateVector(docName);
        let query;
        if (sv) {
            query = format(`
				UPDATE %I
				SET value = %L
				WHERE id = %L
				RETURNING *;`, this.tableName, bufferValue, sv.id);
        }
        else {
            query = format(`
				INSERT INTO %I (docname, value, version)
				VALUES (%L, %L, 'v1_sv')
				RETURNING *;`, this.tableName, docName, bufferValue);
        }
        const res = await this.pool.query(query);
        return res.rows[0];
    }
    async clearUpdatesRange(docName, from, to) {
        const query = format(`
            DELETE FROM %I
            WHERE docname = %L
			AND version = 'v1'
            AND id >= %L
            AND id < %L;`, this.tableName, docName, from, to);
        await this.pool.query(query);
    }
    async readUpdatesAsCursor(docName, callback) {
        let offset = 0;
        const limit = 100;
        let rowsCount = 0;
        let rows = [];
        do {
            const query = format(`
            SELECT * FROM %I
            WHERE docname = %L
            AND version = 'v1'
            ORDER BY id
            LIMIT %L OFFSET %L;`, this.tableName, docName, limit, offset);
            const res = await this.pool.query(query);
            rows = res.rows;
            rowsCount += rows.length;
            callback(rows);
            offset += limit;
        } while (rows.length === limit);
        return rowsCount;
    }
    async deleteDocument(docName) {
        const query = format(`
                DELETE FROM %I
                WHERE docname = %L
                RETURNING *;`, this.tableName, docName);
        const res = await this.pool.query(query);
        return res.rows[0];
    }
    async close() {
        await this.pool.end();
    }
}

exports.PgAdapter = PgAdapter;
